compiler/passes/scopeResolve.cpp

```diff
@@ -2251,13 +2268,27 @@ static bool skipUse(std::map<Symbol*, std::vector<VisibilityStmt*> >* seen,
   std::vector<VisibilityStmt*> vec = (*seen)[useSE->symbol()];
 
   if (vec.size() > 0) {
-    // We've already seen at least one use or import of this module.  Since
-    // imports only impact qualified naming, by definition any previous use or
-    // import of the same module will provide at least as much as this import
-    return true;
+    // We've already seen at least one use or import of this module, but it
+    // might not be thorough enough to justify skipping the newest 'import'
+    for_vector(VisibilityStmt, stmt, vec) {
+      if (UseStmt* use = toUseStmt(stmt)) {
+        if (current->providesNewSymbols(use) == false) {
```

Comment:
Is this whole providesNewSymbols thing an optimization? It feels like an optimization we could find a better way to write (e.g. using a `std::map` or `std::set` somewhere).

Reply:
Kinda.  It's used in buildBreadthFirstModuleList as part of ensuring that we don't loop forever when building the list of modules to traverse and their order.  The presence of any import or use statement that picks and chooses what it enables means that we have to keep track of the ones we've seen until we find a blanket use statement, so we don't accidentally shut ourselves off from certain symbols, allow symbols that shouldn't be found, or allow symbols to be found sooner than they should be.  I'm sure there are other way to determine this, and possibly ways to do it faster, but I do try to organize it to get quickly out of the computation when we can